File: scripts/dev/native-host/hash-package-tree.mjs
#!/usr/bin/env node
// Compute MAINFRAME's canonical SHA-256 identity for an installed package tree.
// This runtime implementation intentionally uses the same Node.js executable
// required by npm-installed coding-agent wrappers, avoiding a Python dependency
// on stock macOS. Keep its byte stream identical to hash-package-tree.py.

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";

const DOMAIN = Buffer.from("MAINFRAME-PACKAGE-TREE-SHA256-V1\0", "utf8");

function die(message) {
  process.stderr.write(`${message}\n`);
  process.exit(1);
}

function utf8Compare(left, right) {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}

function assertRealDirectory(candidate, message) {
  let metadata;
  try {
    metadata = fs.lstatSync(candidate, { bigint: true });
  } catch {
    die(message);
  }
  if (metadata.isSymbolicLink() || !metadata.isDirectory()) {
    die(message);
  }
}

function normalizeSelections(root, values) {
  const normalized = new Set();
  for (const value of values) {
    const canonical = path.posix.normalize(value);
    const parts = value.split("/");
    if (
      !value ||
      path.posix.isAbsolute(value) ||
      canonical !== value ||
      parts.some((part) => !part || part === "." || part === "..")
    ) {
      die(`selection must be a canonical relative path: ${value}`);
    }
    const selected = path.join(root, ...parts);
    assertRealDirectory(
      selected,
      `selected package tree must be a real directory: ${value}`,
    );
    normalized.add(value);
  }

  return [...normalized]
    .filter(
      (value) =>
        ![...normalized].some(
          (other) => other !== value && value.startsWith(`${other}/`),
        ),
    )
    .sort(utf8Compare);
}

function selectionIncludes(relative, selections) {
  if (selections.length === 0) return true;
  return selections.some(
    (selected) =>
      relative === selected ||
      relative.startsWith(`${selected}/`) ||
      selected.startsWith(`${relative}/`),
  );
}

function main() {
  if (process.argv.length < 3) {
    die(
      "usage: hash-package-tree.mjs PACKAGE_DIRECTORY " +
        "[SELECTED_RELATIVE_DIRECTORY ...]",
    );
  }

  let root = process.argv[2];
  assertRealDirectory(root, `package tree must be a real directory: ${root}`);
  root = fs.realpathSync(root);
  const selections = normalizeSelections(root, process.argv.slice(3));
  const entries = [];

  function visit(current, currentRelative) {
    let names;
    try {
      names = fs.readdirSync(current).sort(utf8Compare);
    } catch (error) {
      die(`could not read package tree directory: ${current}: ${error.message}`);
    }

    for (const name of names) {
      const relative = currentRelative ? `${currentRelative}/${name}` : name;
      if (!selectionIncludes(relative, selections)) continue;
      const entryPath = path.join(current, name);
      let metadata;
      try {
        metadata = fs.lstatSync(entryPath, { bigint: true });
      } catch (error) {
        die(`could not inspect package tree entry: ${relative}: ${error.message}`);
      }
      if (metadata.isSymbolicLink()) {
        die(`package tree contains a symbolic link: ${relative}`);
      }
      if (!metadata.isDirectory() && !metadata.isFile()) {
        die(`package tree contains an unsupported entry: ${relative}`);
      }
      entries.push({ relative, entryPath, metadata });
      if (metadata.isDirectory()) visit(entryPath, relative);
    }
  }

  visit(root, "");
  entries.sort((left, right) => utf8Compare(left.relative, right.relative));

  const digest = crypto.createHash("sha256");
  digest.update(DOMAIN);
  const buffer = Buffer.allocUnsafe(1024 * 1024);
  for (const { relative, entryPath, metadata } of entries) {
    const encodedPath = Buffer.from(relative, "utf8");
    if (encodedPath.includes(0)) die(`package tree path contains NUL: ${relative}`);
    if (metadata.isDirectory()) {
      digest.update(Buffer.from("D\0", "utf8"));
      digest.update(encodedPath);
      digest.update(Buffer.from([0]));
      continue;
    }

    const size = metadata.size;
    digest.update(Buffer.from("F\0", "utf8"));
    digest.update(encodedPath);
    digest.update(Buffer.from([0]));
    digest.update(Buffer.from(`${size}\0`, "ascii"));

    let descriptor;
    let bytesRead = 0n;
    try {
      descriptor = fs.openSync(entryPath, "r");
      while (true) {
        const count = fs.readSync(descriptor, buffer, 0, buffer.length, null);
        if (count === 0) break;
        bytesRead += BigInt(count);
        digest.update(buffer.subarray(0, count));
      }
    } catch (error) {
      die(`could not read package tree file: ${relative}: ${error.message}`);
    } finally {
      if (descriptor !== undefined) fs.closeSync(descriptor);
    }
    if (bytesRead !== size) {
      die(
        "package tree file is dataless, truncated, or changed while hashing: " +
          `${relative} (stat size ${size}, bytes read ${bytesRead})`,
      );
    }
  }

  process.stdout.write(`${digest.digest("hex")}\n`);
}

main();

File: scripts/dev/native-host/hash-package-tree.py
#!/usr/bin/env python3
"""Compute MAINFRAME's canonical SHA-256 identity for an installed package tree."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import pathlib
import re
import stat
import sys
from typing import NoReturn


DOMAIN = b"MAINFRAME-PACKAGE-TREE-SHA256-V1\0"


def die(message: str) -> NoReturn:
    raise SystemExit(message)


def normalize_selections(root: pathlib.Path, values: list[str]) -> tuple[str, ...]:
    """Return minimal, validated POSIX-relative directory selections."""

    normalized: set[str] = set()
    for value in values:
        candidate = pathlib.PurePosixPath(value)
        if (
            not value
            or candidate.is_absolute()
            or value != candidate.as_posix()
            or any(part in ("", ".", "..") for part in candidate.parts)
        ):
            die(f"selection must be a canonical relative path: {value}")
        selected = root.joinpath(*candidate.parts)
        if selected.is_symlink() or not selected.is_dir():
            die(f"selected package tree must be a real directory: {value}")
        normalized.add(candidate.as_posix())

    # Selecting a parent already includes every descendant selection.
    minimal = {
        value
        for value in normalized
        if not any(value.startswith(other + "/") for other in normalized if other != value)
    }
    return tuple(sorted(minimal))


def selection_includes(relative: str, selections: tuple[str, ...]) -> bool:
    if not selections:
        return True
    return any(
        relative == selected
        or relative.startswith(selected + "/")
        or selected.startswith(relative + "/")
        for selected in selections
    )


def collect_entries(
    root: pathlib.Path, selections: tuple[str, ...]
) -> list[tuple[str, pathlib.Path, int, int]]:
    entries: list[tuple[str, pathlib.Path, int, int]] = []
    for current, directory_names, file_names in os.walk(
        root, topdown=True, followlinks=False
    ):
        directory_names.sort()
        file_names.sort()
        current_path = pathlib.Path(current)
        current_relative = current_path.relative_to(root).as_posix()
        if current_relative == ".":
            current_relative = ""
        directory_names[:] = [
            name
            for name in directory_names
            if selection_includes(
                f"{current_relative}/{name}" if current_relative else name,
                selections,
            )
        ]
        file_names = [
            name
            for name in file_names
            if selection_includes(
                f"{current_relative}/{name}" if current_relative else name,
                selections,
            )
        ]
        for name in directory_names + file_names:
            path = current_path / name
            metadata = path.lstat()
            relative = path.relative_to(root).as_posix()
            if stat.S_ISLNK(metadata.st_mode):
                die(f"package tree contains a symbolic link: {relative}")
            if not (stat.S_ISDIR(metadata.st_mode) or stat.S_ISREG(metadata.st_mode)):
                die(f"package tree contains an unsupported entry: {relative}")
            entries.append((relative, path, metadata.st_mode, metadata.st_size))
    return entries


def hash_entries(entries: list[tuple[str, pathlib.Path, int, int]]) -> str:
    """Hash the canonical typed path inventory and every regular-file byte."""

    digest = hashlib.sha256(DOMAIN)
    for relative, path, mode, _observed_size in sorted(entries):
        encoded_path = relative.encode("utf-8")
        if b"\0" in encoded_path:
            die(f"package tree path contains NUL: {relative}")
        if stat.S_ISDIR(mode):
            digest.update(b"D\0" + encoded_path + b"\0")
            continue

        size = path.stat().st_size
        digest.update(b"F\0" + encoded_path + b"\0")
        digest.update(str(size).encode("ascii") + b"\0")
        bytes_read = 0
        with path.open("rb") as source:
            while chunk := source.read(1024 * 1024):
                bytes_read += len(chunk)
                digest.update(chunk)
        if bytes_read != size:
            die(
                "package tree file is dataless, truncated, or changed while "
                f"hashing: {relative} (stat size {size}, bytes read {bytes_read})"
            )

    return digest.hexdigest()


def print_inventory(entries: list[tuple[str, pathlib.Path, int, int]]) -> None:
    """Emit an injection-safe inventory without reading file contents."""

    print("selected installed-tree inventory (D=directory, F=regular file):", file=sys.stderr)
    for relative, _path, mode, observed_size in sorted(entries):
        encoded_relative = json.dumps(relative, ensure_ascii=True)
        if stat.S_ISDIR(mode):
            print(f"  D {encoded_relative}", file=sys.stderr)
        else:
            print(f"  F {encoded_relative} size={observed_size}", file=sys.stderr)


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Compute MAINFRAME's canonical installed package-tree identity."
    )
    parser.add_argument(
        "--expected",
        metavar="SHA256",
        help="fail with installed-tree diagnostics unless the digest matches SHA256",
    )
    parser.add_argument(
        "--installed-label",
        metavar="LABEL",
        help="host/package label used by --expected diagnostics",
    )
    parser.add_argument("package_directory", metavar="PACKAGE_DIRECTORY")
    parser.add_argument(
        "selections",
        metavar="SELECTED_RELATIVE_DIRECTORY",
        nargs="*",
    )
    arguments = parser.parse_args()
    if (arguments.expected is None) != (arguments.installed_label is None):
        parser.error("--expected and --installed-label must be supplied together")
    if arguments.expected is not None and not re.fullmatch(
        r"[0-9a-f]{64}", arguments.expected
    ):
        parser.error("--expected must be a lowercase SHA-256 digest")
    if arguments.installed_label is not None and not re.fullmatch(
        r"[A-Za-z0-9][A-Za-z0-9._+ -]{0,79}", arguments.installed_label
    ):
        parser.error("--installed-label contains unsafe characters")
    return arguments


def main() -> None:
    arguments = parse_arguments()
    entries: list[tuple[str, pathlib.Path, int, int]] = []
    try:
        root = pathlib.Path(arguments.package_directory)
        if root.is_symlink() or not root.is_dir():
            die(f"package tree must be a real directory: {root}")
        root = root.resolve(strict=True)
        selections = normalize_selections(root, arguments.selections)
        entries = collect_entries(root, selections)
        digest = hash_entries(entries)
    except SystemExit as error:
        if arguments.expected is None:
            raise
        print(
            f"ERROR: installed {arguments.installed_label} tree could not be hashed; "
            "installed-tree contamination or incomplete package data",
            file=sys.stderr,
        )
        if str(error):
            encoded_error = json.dumps(str(error), ensure_ascii=True)
            print(f"tree-hash error: {encoded_error}", file=sys.stderr)
        if entries:
            print_inventory(entries)
        raise SystemExit(1) from None

    if arguments.expected is not None and digest != arguments.expected:
        print(
            f"ERROR: installed {arguments.installed_label} tree does not match host "
            "manifest; installed-tree contamination or incomplete package data",
            file=sys.stderr,
        )
        print(f"expected selected-tree SHA-256: {arguments.expected}", file=sys.stderr)
        print(f"actual selected-tree SHA-256:   {digest}", file=sys.stderr)
        print_inventory(entries)
        raise SystemExit(1)

    print(digest)


if __name__ == "__main__":
    main()
